### DISCUSSION TITLE
Withdraw on behalf of

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#39] [Ideas]
In an offline discussion, @razgraf made the suggestion to implement a "withdraw on behalf of" functionality, which would enable  third-party relayers to pay the gas costs for withdrawing for end-users.

My only reservation is that it would take quite some time to implement this, and we already have a long list of more urgent matters to tackle. Multiple new functions and one big mapping would be needed to add support for withdrawing on behalf of other users.

That said, there are two alternative implementations that would achieve the same goal, albeit they would involve different assumptions:

1. What if we let anyone call the `withdraw` function? Obviously the concern with this is that anyone would be able to trigger taxable events for other people.
2. In a similar vein to #1, what if we add a new boolean property in the `Stream` struct that by default is set to `false` but users can set to `true` which would allow anyone to call the `withdraw` function?

**Comment by razgraf** [comment]
There is a 3rd option of having a special allowance system, similar to the management of the `funders`.

I'll bring yet another reason for which this functionality is very important. For our custom vault/routers/erc4626 and even as per the example given in #41 we will have the possibility of hiding certain actions from the user. Example: they'll be able to stream native tokens while we do the wrapping on create and the **unwrapping on withdraw**. In order to be able to do that, the user won't call the `withdraw` method of the sablier contract but the `withdraw/claim` method of the vault, which will perform the withdrawal on their behalf, unwrap the token and send it to the user in the end.

**Reply by PaulRBerg** [reply]
Your 3rd option is my 1st option, which I implicitly referred to when I said "Multiple new functions and one big mapping would be needed".

> In order to be able to do that, the user won't call the `withdraw` method of the sablier contract but the `withdraw/claim` method of the vault, which will perform the withdrawal on their behalf, unwrap the token and send it to the user in the end.

I don't think that this is connected to the said discussion about the "withdraw on behalf of" functionality. If a user calls the `withdraw` method on the vault, and the vault calls `withdrawOnBehalfOf`, the tokens would be transferred to the user, thus leaving the vault contract no ability to unwrap the tokens, or for that matter any post-transfer action. The vault would have to be the recipient of the stream, but this is a different idea to discuss.

**Reply by razgraf** [reply]
Hmm I was thinking more like:
1. User allows vault to spend `asset` ~~and `wAsset`~~
2. Stream gets created - user calls vault method `create` which takes `asset`, wraps into `wAsset` and calls `sablier.create()`
3. Stream happens
4. User wants to withdraw - user calls vault method `withdraw` [the following happen in the same transaction]
 4.1. Vault calls `sablier.withdrawOnBehalfTo()` - it withdraws `wAsset` and transfers it to itself (I agree this introduces a new method and a strong attack vector)
 4.2. Vault has the `wAsset` so it proceeds to unwrap it into `asset`
 4.3. Vault sends the `asset` to user
 
LE: Actually if user allows vault to spend `wAsset`, we might not even need the `onBehalfTo` variant
 
I also like the idea of the vault being the recipient. Even better we could think of some case where streaming vault tokens or special-logic tokens is *only* done through a proxy/user.

I guess we need to have a session brainstorming just these vault use-cases to make sure we do it right.

**Reply by PaulRBerg** [reply]
Re point no. 1 - native assets like ETH don't have any allowance mechanism.

Re other points - this is a different feature proposal. I thought we were discussing **withdraw on behalf OF**, not **withdraw on behalf TO**. We can discuss this idea as well, but again, we have to weigh it in compared to the other more urgent (and necessary) tasks on our plate.

> LE: Actually if user allows vault to spend `wAsset`, we might not even need the `onBehalfTo` variant

We would still need it, due to the reasons given in my first comment.

> I also like the idea of the vault being the recipient.

Yes but it's not practical. Senders would have to be aware of this and ask recipients to create some specialized proxy/ vault. Very cumbersome.

> I guess we need to have a session brainstorming just these vault use-cases to make sure we do it right.

Yup.

**Reply by razgraf** [reply]
Re point no. 1 - I attempted to solve the wrap/unwrap use-case globally, be it for ETH/WETH or DAI/CDAI, the latter being a type of wrapping as well. Hence the allowance.

**Reply by PaulRBerg** [reply]
Yes, for DAI/ CDAI there would have to be a wrapping step, but you cannot use the same implementation for ETH/ WETH. I mean, you could, by adding if statements, but I don't see why we would do that when we can just have different periphery contracts for native assets vs yield-bearing assets.

**Reply by razgraf** [reply]
Well that's exactly what we should do, split it in 2 different periphery! But the logic remains more or less the same (except allowances and payable methods).

**Comment by razgraf** [comment]
As asked I'll share my thoughts on a public `withdraw` function.

**Pro**
- Anyone can trigger a withdrawal, thus being able to subsidize gas costs
- Lower gas costs due to the lack of caller checks

**Con**
- Anyone could trigger a tax event for the recipient by calling withdraw
- The freedom of choice is out of the hands of the recipient - we would break the **principle of least privilege**

---
I still think we would benefit from creating an access control scheme where the user can nominate some entities that handle withdrawal on their behalf. The natural question is if we should implement this privilege manager for:

1. `withdraw` direct to recipient - `onlyRecipient, onlySender(?), onlyAuthorized`
2. `withdrawTo` redirect funds to any address - `onlyRecipient, onlySender(?), onlyAuthorized`


**Pro**
- Only authorized actors can trigger a tax event
- Withdrawals could be safely automated and redirected (imagine a relayer doing DCAs on your behalf)
- Withdrawals could be combined (zaps) with other transactions *maybe* even without the need for a proxy contract

**Con**
- Increases complexity and gas costs

**Reply by PaulRBerg** [reply]
I assume that what you meant here by the access control scheme is a design whereby the user would allow a third-party to withdraw on their behalf for *any* stream? That'd be interesting, for sure!

**Comment by razgraf** [comment]
Another argument against a public `withdraw` method is the complexity it brings to integrations.
Imagine a collateralization protocol, using a wrapper contract for our NFT streams. As someone randomly calls withdraw, now the balances are split between 
1. what's still locked (and available) in the stream itself 
2. what's now assigned as balance to the wrapper contract (the recipient at the time of withdrawal) - tokens that if not careful could be locked

**Reply by PaulRBerg** [reply]
This is a weak argument, because it is already the case that the sender can withdraw on behalf of the recipient, so the said protocol would have to account for this edge case nonetheless.

**Reply by razgraf** [reply]
Or we allow just the recipient to withdraw. What was the case again for the sender being able to call withdraw?
(The argument above was under the assumption that we completely limit the ability to withdraw to only the recipient of the stream).

**Reply by razgraf** [reply]
Note: my reluctance around this method being public is that giving to much freedom to it could come back and bite us. Heavily restricting (only recipient) or having some sort of constraints feels safer. (e.g. handling an access-control list even at proxy level if not the core protocol)

**Reply by PaulRBerg** [reply]
> Or we allow just the recipient to withdraw.

This is a different idea altogether which we should create a bespoke discussion for.

> What was the case again for the sender being able to call withdraw?

This is a very important functionality for the following cases:

~~1. Recipients lose access to their original wallets~~
2. Recipients give senders a CEX address

These are also non-problems if we make the withdraw function public.

**Reply by PaulRBerg** [reply]
> my reluctance around this method being public is that giving to much freedom to it could come back and bite us.

The only problem I can think is *some* users complaining that a tax event was triggered for them. But that is contingent upon a specific interpretation of Sablier streams for tax purposes, so in other interpretations withdrawals need not be tax events at all.

> having some sort of constraints feels safer

"safety" should be defined in terms of risk of funds being lost to third-party hackers. I agree that allowing anyone to call `withdraw` increases the risk a little bit, but just like any other powerful functionality in Sablier, we will test it thoroughly.

**Reply by razgraf** [reply]
Quick thoughts on your reply above:

1. if users were to lose access to their wallet, there's no way to save funds: those would stay locked either in the contract or in the user's wallet as they wouldn t be able to sign a transfer out anyway.
2. cex is a good reason, but for a fee they can get the funds out + sender could cancel (?) and trigger this transfer

**Reply by razgraf** [reply]
Actually could you draft a list of pros (and cons even, if you re kind enough to gather them) of why `withdraw` should or should not be public

--
We can discuss levels of access later too (sender v. recipient only, especially cause with a call to the recipient's proxy + an allow_list they could give access to exactly who they want to)

**Reply by PaulRBerg** [reply]
1. That's correct.
2. Yup, they could cancel too, but I would move this point to another discussion. Don't know what you mean by "for a fee they can get the funds out" though.

**Reply by maxdesalle** [reply]
Maybe a stupid question, but wouldn't be possible to make it so that users can "whitelist" an address, to allow that address to call the withdraw function on their behalf?

**Reply by PaulRBerg** [reply]
@maxdesalle yep, that would be possible, but my argument is that instead of asking users to manually whitelist a relayer, why not allow any relayer to do this, by default, globally?

Significantly improved UX.

**Reply by PaulRBerg** [reply]
> Actually could you draft a list of pros (and cons even, if you re kind enough to gather them) of why `withdraw` should or should not be public

Will do, @razgraf, in another discussion.

**Reply by maxdesalle** [reply]
@paulrberg users should always be in control IMO, even if it's at the cost of UX.

Even if it doesn't make sense from a rational standpoint, I wouldn't want someone being able to transfer my own funds without my approval. Psychologically, users want to feel safe and in control, which I don't think we can do unless we explicitly ask for their approval first.

**Reply by razgraf** [reply]
@maxdesalle totally, we have some discussions on `withdrawOnBehalf` which would require an allow-list system.

It's just a matter of added complexity in the mplementation, which I can def. see although I'm a fan of the solution.

**Reply by PaulRBerg** [reply]
> Even if it doesn't make sense from a rational standpoint, I wouldn't want someone being able to transfer my own funds without my approval.

As I had said above, this assumes a specific interpretation of tax events in Sablier.

If you consider the funds in Sablier as already yours, there wouldn't be anyone "transferring" your funds, because they are already yours.

> It's just a matter of added complexity in the mplementation, which I can def. see although I'm a fan of the solution.

The complexity of the implementation is just the cherry on top - my focus is the UX.

**Reply by razgraf** [reply]
I's argue the better UX here (for both a user and a dev) is to be in charge of who can do what.

**Reply by PaulRBerg** [reply]
@razgraf what it means for a user to "be in charge" is contingent upon how a user interprets the "moneyness" of their Sablier streams.

**Reply by maxdesalle** [reply]
It's not even about taxes @paulrberg, it's just that I wouldn't want a random stranger being able to withdraw funds from Sablier to my wallet without my approval, just like I wouldn't want a random stranger being able to transfer funds between two bank accounts in my control.

Yes, the money stays mine, but it's still an invasion of personal boundaries.

**Reply by maxdesalle** [reply]
Worth noting that a lot of recipients in Sablier are people who are really not technical when it comes to crypto. It's usually the stream creators who are technical (company founder), streaming to recipients who are sometimes technical, but sometimes not (company employees).

If someone where to call the `withdraw` function on their behalf, without them knowing, chances are high non-technical recipients would believe they are hacked.

**Reply by PaulRBerg** [reply]
@maxdesalle The year is 2025 and Sablier is completely integrated in MetaMask, in the sense that MetaMask tracks all of your streams and shows up your total balance across all of your streams. Suppose then that you want to make a transfer - if the actual balance in your EOA is not sufficient, MetaMask would simply call `withdraw` on as many streams as needed, so you can then perform your transfer.

The only reason you know that "a random stranger" can perform an action is a matter of web3 wallets UX - that is, how they display your balance with respect to Sablier (hint: they don't do it at all today).

I will expound on this point in another discussion because it became difficult to track all the arguments that have been made in this comment thread.

**Reply by PaulRBerg** [reply]
> If someone where to call the `withdraw` function on their behalf, without them knowing, chances are high non-technical recipients would believe they are hacked.

Why would a hacker pay you? If they wanted to hack you, they would have taken money from you.

**Comment by PaulRBerg** [comment]
This proposal has been morphed into https://github.com/sablierhq/v2-core/discussions/155. Locking.
